### PR TITLE
Rewrite `pin` utility.

### PR DESCRIPTION
**Pull request by Kakadu**
More error messages. More int conversion checking. @Deryugin

### CODE REVIEW

File: src/cmds/hardware/pin/pin.c [hunk 1]
```diff
@@ -10,71 +10,126 @@
...
-		return GPIO_PIN_13;
-	case 14:
-		return GPIO_PIN_14;
-	case 15:
-		return GPIO_PIN_15;
-	default:
+	char *end;
+	int n = strtol(name, &end, 10);
+	if (*end == '\0')
+		return pin_index_of_int(n);
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: Please add braces around this line for code style consistency

File: src/cmds/hardware/pin/pin.c [hunk 2]
```diff
@@ -10,71 +10,126 @@
...
+ *             If function fails result value is not specified.
+ * @return TRUE if functions successfully parsed GPIO and (maybe) pin index
+ *         from input string. On fail returns FALSE.
+ */
+static int gpio_by_name(char *name, void **gpio, gpio_mask_t *pin) {
+	if (!strncmp("GPIO", name, 4) || !strncmp("gpio", name, 4))
+		name += 4;
+
+	if (strlen(name) == 0)
 		return 0;
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: Add another pair of braces..

File: src/cmds/hardware/pin/pin.c [hunk 3]
```diff
@@ -10,71 +10,126 @@
...
+ *             pin. If pin is not specified contains PIN_UNSPECIFIED.
+ *             If function fails result value is not specified.
+ * @return TRUE if functions successfully parsed GPIO and (maybe) pin index
+ *         from input string. On fail returns FALSE.
+ */
+static int gpio_by_name(char *name, void **gpio, gpio_mask_t *pin) {
+	if (!strncmp("GPIO", name, 4) || !strncmp("gpio", name, 4))
+		name += 4;
+
+	if (strlen(name) == 0)
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: Add another pair of braces..

File: src/cmds/hardware/pin/pin.c [hunk 4]
```diff
@@ -10,71 +10,126 @@
...
+ * @param name Input string to parse
+ * @param gpio Output parameter. If function returns successfully contains
+ *             information about specified GPIO.
+ * @param pin  Output parameter. If pin is specified contains index of specified
+ *             pin. If pin is not specified contains PIN_UNSPECIFIED.
+ *             If function fails result value is not specified.
+ * @return TRUE if functions successfully parsed GPIO and (maybe) pin index
+ *         from input string. On fail returns FALSE.
+ */
+static int gpio_by_name(char *name, void **gpio, gpio_mask_t *pin) {
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: I think it's a good idea to return pointer to GPIO address and get rid of **gpio argument, this way code will be shorter and more understandable.

File: src/cmds/hardware/pin/pin.c [hunk 5]
```diff
@@ -10,71 +10,126 @@
...
+	case 'e':	case 'E': *gpio = GPIOE; return 1;
+	case 'f':	case 'F': *gpio = GPIOF; return 1;
 	}
+
+	fprintf(stderr, "Can't convert character '%c' to GPIO. Error.", name[0]);
+	return 0;
+}
+
+#define LEVEL_SET 1
+#define LEVEL_UNSET 0
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: You can use enum GPIO_PinState from stm32f4xx_hal_gpio.h

File: src/cmds/hardware/pin/pin.c [hunk 6]
```diff
@@ -10,71 +10,126 @@
...
+
+int extract_level(int pos, int argc, char **argv) {
+	if (pos >= argc)
+		// we suppose default inspecified value is `set`
+		return LEVEL_SET;
```
Comment (Deryugin) [inline src/cmds/hardware/pin/pin.c]: Another pair of braces missing

File: src/cmds/hardware/pin/pin.c (hunk 5 above)
Comment (Kakadu) [inline src/cmds/hardware/pin/pin.c]: Yeah, but will it make this utility available only on STMs?

Comment (Kakadu) [inline src/cmds/hardware/pin/pin.c]: Ah, it seems that it is already STM-only.

**Comment by Kakadu** [comment]
issues seems to be fixed but build bot has some connectivity issues so build didn't pass successfully.

**Comment by anton-bondarev** [comment]
@Kakadu Now that's all right

**Comment by anton-bondarev** [comment]
LGTM

**Pull request merged**